plugins/LanguagesManager/LanguagesManager.php

```diff
@@ -137,18 +128,46 @@ public function uninstall()
     }
 
     /**
+     * First check forced by request language setting,
+     * then - session language setting, then - from browser,
+     * and finally default configured setting.
+     *
      * @return string Two letters language code, eg. "fr"
      */
-    public static function getLanguageCodeForCurrentUser()
+    public static function getCurrentLanguage()
     {
-        $languageCode = self::getLanguageFromPreferences();
-        if (!API::getInstance()->isLanguageAvailable($languageCode)) {
-            $languageCode = Common::extractLanguageCodeFromBrowserLanguage(Common::getBrowserLanguage(), API::getInstance()->getAvailableLanguages());
+        $languageCodes = array(
+            Common::getRequestVar('language', ''),
+            self::getLanguageForSession(),
+            self::getLanguageCodeForCurrentUser(),
+            StaticContainer::get('Piwik\Translation\Translator')->getDefaultLanguage(),
+        );
+        foreach ($languageCodes as $languageCode) {
+            if (!empty($languageCode) && API::getInstance()->isLanguageAvailable($languageCode)) {
+                return $languageCode;
+            }
         }
-        if (!API::getInstance()->isLanguageAvailable($languageCode)) {
-            $languageCode = Translate::getLanguageDefault();
+        return null;
+    }
+
+    /**
+     * @return string Two letters language code, eg. "fr"
+     */
+    public static function getLanguageCodeForCurrentUser()
+    {
+        $languageCodes = array(
+            self::getLanguageFromPreferences(),
+            Common::extractLanguageCodeFromBrowserLanguage(
+                Common::getBrowserLanguage(),
+                API::getInstance()->getAvailableLanguages()
+            ),
+        );
+        foreach ($languageCodes as $languageCode) {
+            if (!empty($languageCode) && API::getInstance()->isLanguageAvailable($languageCode)) {
+                return $languageCode;
+            }
         }
-        return $languageCode;
+        return null;
```

Comment:
I think because of the refactoring you did the logic has changed:

- the method can return null, whereas before it would return `self::getLanguageFromPreferences()` in the worst case
- the priority of many sources is inversed, so it seems to me that it will always return `self::getLanguageFromPreferences()` in any case (and ignore the rest)